File: backend/src/patterns/observer/EventSubject.ts
import { Event } from '../../types';
import { RegistrationRepository } from '../../repositories/RegistrationRepository';
import {
  NotificationBridge,
  EventUpdatedNotification,
  EventCancelledNotification,
  OrganizerAnnouncementNotification,
  EmailChannel,
  InAppChannel
} from '../bridge/NotificationBridge';

/**
 * OBSERVER DESIGN PATTERN
 * 
 * Problem: When an event venue, date, schedule, or status changes, manually writing logic
 * to find and notify every student, organizer, and administrator leads to tight coupling and duplicated logic.
 * 
 * Solution: Define a one-to-many dependency between EventSubject (Subject) and Observer objects
 * so that when the event changes, all dependents are notified automatically.
 */

export interface IEventObserver {
  getObserverName(): string;
  onEventChanged(event: Event, changeType: string, message?: string): Promise<void>;
}

export class EventSubject {
  private static instance: EventSubject | null = null;
  private observers: IEventObserver[] = [];

  private constructor() {}

  public static getInstance(): EventSubject {
    if (!EventSubject.instance) {
      EventSubject.instance = new EventSubject();
    }
    return EventSubject.instance;
  }

  public attach(observer: IEventObserver): void {
    if (!this.observers.some(o => o.getObserverName() === observer.getObserverName())) {
      this.observers.push(observer);
    }
  }

  public detach(observer: IEventObserver): void {
    this.observers = this.observers.filter(o => o.getObserverName() !== observer.getObserverName());
  }

  public async notifyObservers(event: Event, changeType: string, message?: string): Promise<void> {
    for (const observer of this.observers) {
      try {
        await observer.onEventChanged(event, changeType, message);
      } catch (err) {
        console.error(`Observer error [${observer.getObserverName()}]:`, err);
      }
    }
  }
}

// ----------------------------------------------------
// Concrete Observers
// ----------------------------------------------------

export class RegisteredStudentsObserver implements IEventObserver {
  private regRepo = new RegistrationRepository();

  getObserverName(): string { return 'RegisteredStudentsObserver'; }

  async onEventChanged(event: Event, changeType: string, message?: string): Promise<void> {
    const registrations = await this.regRepo.findByEvent(event.eventId);
    if (registrations.length === 0) return;

    let notificationBridge: NotificationBridge;

    if (changeType === 'CANCELLED') {
      notificationBridge = new EventCancelledNotification(new EmailChannel());
    } else if (changeType === 'ANNOUNCEMENT') {
      notificationBridge = new OrganizerAnnouncementNotification(new InAppChannel());
    } else {
      notificationBridge = new EventUpdatedNotification(new InAppChannel());
    }

    for (const reg of registrations) {
      if (reg.status === 'CONFIRMED' || reg.status === 'PENDING') {
        await notificationBridge.dispatch(reg.studentId, event.eventId, event.title, message);
      }
    }
  }
}

export class OrganizerObserver implements IEventObserver {
  getObserverName(): string { return 'OrganizerObserver'; }

  async onEventChanged(event: Event, changeType: string, message?: string): Promise<void> {
    const inAppBridge = new EventUpdatedNotification(new InAppChannel());
    await inAppBridge.dispatch(
      event.organizerId,
      event.eventId,
      event.title,
      `[Organizer Confirmation] Action '${changeType}' recorded for your event "${event.title}".`
    );
  }
}

export class AdminAuditObserver implements IEventObserver {
  getObserverName(): string { return 'AdminAuditObserver'; }

  async onEventChanged(event: Event, changeType: string, message?: string): Promise<void> {
    console.log(`[Admin Audit Observer] Event #${event.eventId} ("${event.title}") triggered change '${changeType}'. Info: ${message || 'None'}`);
  }
}

File: backend-java/src/main/java/com/eventbridge/patterns/observer/EventSubject.java
package com.eventbridge.patterns.observer;

import com.eventbridge.model.Event;
import com.eventbridge.patterns.bridge.EmailChannel;
import com.eventbridge.patterns.bridge.InAppChannel;
import com.eventbridge.patterns.bridge.NotificationBridge;
import com.eventbridge.patterns.bridge.RegistrationConfirmationNotification;
import java.util.ArrayList;
import java.util.List;

/**
 * OBSERVER DESIGN PATTERN
 * 
 * Subject broadcasting event state changes to registered observers.
 */
public class EventSubject {
    private final List<IEventObserver> observers = new ArrayList<>();

    public EventSubject() {
        // Register default observers
        observers.add(new RegisteredStudentsObserver());
        observers.add(new OrganizerObserver());
        observers.add(new AdminAuditObserver());
    }

    public void attach(IEventObserver observer) {
        observers.add(observer);
    }

    public void detach(IEventObserver observer) {
        observers.remove(observer);
    }

    public void notifyObservers(Event event, String changeType, String message) {
        for (IEventObserver obs : observers) {
            obs.onEventChanged(event, changeType, message);
        }
    }
}

interface IEventObserver {
    void onEventChanged(Event event, String changeType, String message);
}

class RegisteredStudentsObserver implements IEventObserver {
    @Override
    public void onEventChanged(Event event, String changeType, String message) {
        System.out.println("📢 [RegisteredStudentsObserver] Event #" + event.getEventId() + " (" + event.getTitle() + ") changed: " + changeType + ". Broadcast message: " + message);
    }
}

class OrganizerObserver implements IEventObserver {
    @Override
    public void onEventChanged(Event event, String changeType, String message) {
        System.out.println("📋 [OrganizerObserver] Event organizer notified of state change for " + event.getTitle());
    }
}

class AdminAuditObserver implements IEventObserver {
    @Override
    public void onEventChanged(Event event, String changeType, String message) {
        System.out.println("🛡️ [AdminAuditObserver] Audit Log entry recorded for Event #" + event.getEventId() + " - Action: " + changeType);
    }
}
